### PR TITLE
Add computed data to VDC storage profile data source

### PR DESCRIPTION
**Pull request by vbauzysvmware**
ref: https://github.com/vmware/terraform-provider-vcd/issues/692, https://github.com/vmware/terraform-provider-vcd/issues/605
dep: https://github.com/vmware/go-vcloud-director/pull/435

Added additional information returned from API to be written to state.

Additional fields:
- limit
- used_storage
-  default
- enabled
- iops_allocated
- units
- iops_settings

IOPS settings details:

- iops_limiting_enabled
- maximum_disk_iops
- default_disk_iops 
- disk_iops_per_gb_max
- iops_limit

### CODE REVIEW

File: .changes/v3.6.0/782-improvements.md [hunk 1]
```diff
@@ -0,0 +1 @@
+* `data/vcd_storage_profile` add IOPS settings and other data to state [GH-782]
```
Comment (lvirbalas) [inline .changes/v3.6.0/782-improvements.md:1]: Please list specific fields that were added - someone may benefit of that.

File: vcd/datasource_vcd_storage_profile.go [hunk 2]
```diff
@@ -32,6 +32,70 @@ func datasourceVcdStorageProfile() *schema.Resource {
...
+			},
+			"default": &schema.Schema{
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "True if this is default storage profile for this vDC. The default storage profile is used when an object that can specify a storage profile is created with no storage profile specified",
+			},
+			"enabled": &schema.Schema{
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "True if this storage profile is enabled for use in the vDC",
```
Comment (lvirbalas) [inline vcd/datasource_vcd_storage_profile.go:53]: Let's use `VDC` in all the messages of the PR.

**Review by lvirbalas — CHANGES_REQUESTED** [review]

File: vcd/resource_vcd_vapp_vm.go [hunk 3]
```diff
@@ -653,6 +653,9 @@ func genericResourceVmCreate(d *schema.ResourceData, meta interface{}, vmType ty
 				vmComputePolicy = nil
 			}
 			vmTemplate := vmTemplatefromVappTemplate(d.Get("vm_name_in_template").(string), vappTemplate.VAppTemplate)
+			if vmTemplate == nil {
+				return fmt.Errorf("[VM creation] vm template isn't found. Please check vApp templat %s : %s", vmName, err)
```
Comment (adambarreiro) [inline vcd/resource_vcd_vapp_vm.go:657]:
```suggestion
				return fmt.Errorf("[VM creation] VM template isn't found. Please check vApp template %s : %s", vmName, err)
```

**Review by adambarreiro — APPROVED** [review]

File: vcd/datasource_vcd_storage_profile.go (hunk 2 above)
Comment (vbauzysvmware) [reply vcd/datasource_vcd_storage_profile.go]: fixed

File: .changes/v3.6.0/782-improvements.md (hunk 1 above)
Comment (vbauzysvmware) [reply .changes/v3.6.0/782-improvements.md]: added

File: vcd/resource_vcd_vapp_vm.go (hunk 3 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_vapp_vm.go:657]: fixed

File: .changes/v3.6.0/782-improvements.md (hunk 1 above)
Comment (lvirbalas) [reply .changes/v3.6.0/782-improvements.md]: Thanks. I want that message to be concrete though: is there a reason to have the two words "like", can they be removed?

File: vcd/datasource_vcd_storage_profile.go [hunk 4]
```diff
@@ -32,6 +32,70 @@ func datasourceVcdStorageProfile() *schema.Resource {
 				Required:    true,
 				Description: "Name of storage profile",
 			},
+			"limit": &schema.Schema{
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "Maximum number of storage bytes (scaled by Units) allocated for this profile. A value of 0 is understood to mean `maximum possible`",
```
Comment (Didainius) [inline vcd/datasource_vcd_storage_profile.go:38]: A reference to `Units` field as reading it is hard to understands.

```suggestion
				Description: "Maximum number of storage bytes (scaled by 'units' field) allocated for this profile. A value of 0 is understood to mean `maximum possible`",
```

Comment (Didainius) [inline vcd/datasource_vcd_storage_profile.go:38]:
```suggestion
				Description: "Maximum number of storage bytes (scaled by Units) allocated for this profile. `0` means `maximum possible`",
```

File: vcd/datasource_vcd_storage_profile.go [hunk 5]
```diff
@@ -32,6 +32,70 @@ func datasourceVcdStorageProfile() *schema.Resource {
...
+			},
+			"units": &schema.Schema{
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Scale used to define Limit",
+			},
+			"iops_settings": &schema.Schema{
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "IP ranges used for static pool allocation in the network",
```
Comment (Didainius) [inline vcd/datasource_vcd_storage_profile.go:68]: These are not IP ranges

File: website/docs/d/storage_profile.html.markdown [hunk 6]
```diff
@@ -32,5 +32,25 @@ The following arguments are supported:
...
+<a id="iopsSettings"></a>
+## IOPS settings
+
+(Supported from VCD *v10.2*)
+
+* `iops_limiting_enabled` - True if this storage profile is IOPs-based placement enabled
+* `maximum_disk_iops` - The maximum IOPS value that this storage profile is permitted to deliver. Value of 0 means this max setting is disabled and there is no max disk IOPS restriction
+* `default_disk_iops` - Value of 0 for disk iops means that no iops would be reserved or provisioned for that virtual disk
+* `disk_iops_per_gb_max` - The maximum disk IOPs per GB value that this storage profile is permitted to deliver. A value of 0 means there is no perGB IOPS restriction
+* `iops_limit` - Maximum number of IOPs that can be allocated for this profile. A value of 0 is understood to mean `maximum possible`
```
Comment (Didainius) [inline website/docs/d/storage_profile.html.markdown:56]: Maybe a simplified name

```suggestion
* `iops_limit` - Maximum number of IOPs that can be allocated for this profile. `0` means `maximum possible`
```

File: website/docs/d/storage_profile.html.markdown [hunk 7]
```diff
@@ -32,5 +32,25 @@ The following arguments are supported:
...
+
+<a id="iopsSettings"></a>
+## IOPS settings
+
+(Supported from VCD *v10.2*)
+
+* `iops_limiting_enabled` - True if this storage profile is IOPs-based placement enabled
+* `maximum_disk_iops` - The maximum IOPS value that this storage profile is permitted to deliver. Value of 0 means this max setting is disabled and there is no max disk IOPS restriction
+* `default_disk_iops` - Value of 0 for disk iops means that no iops would be reserved or provisioned for that virtual disk
+* `disk_iops_per_gb_max` - The maximum disk IOPs per GB value that this storage profile is permitted to deliver. A value of 0 means there is no perGB IOPS restriction
```
Comment (Didainius) [inline website/docs/d/storage_profile.html.markdown:55]:
```suggestion
* `disk_iops_per_gb_max` - The maximum disk IOPs per GB value that this storage profile is permitted to deliver. A value of 0 means there is no per GB IOPS restriction
```

File: website/docs/d/storage_profile.html.markdown [hunk 8]
```diff
@@ -32,5 +32,25 @@ The following arguments are supported:
...
+* `iops_settings` - A block providing iops settings. See [IOPS settings](#iopsSettings) below for details.
+
+<a id="iopsSettings"></a>
+## IOPS settings
+
+(Supported from VCD *v10.2*)
+
+* `iops_limiting_enabled` - True if this storage profile is IOPs-based placement enabled
+* `maximum_disk_iops` - The maximum IOPS value that this storage profile is permitted to deliver. Value of 0 means this max setting is disabled and there is no max disk IOPS restriction
+* `default_disk_iops` - Value of 0 for disk iops means that no iops would be reserved or provisioned for that virtual disk
```
Comment (Didainius) [inline website/docs/d/storage_profile.html.markdown:54]:
```suggestion
* `default_disk_iops` - Value of 0 for disk iops means that no IOPS would be reserved or provisioned for that virtual disk
```

File: website/docs/d/storage_profile.html.markdown [hunk 9]
```diff
@@ -32,5 +32,25 @@ The following arguments are supported:
...
+* `iops_allocated` - Total IOPS currently allocated to this storage profile
+* `units` - Scale used to define Limit
+* `iops_settings` - A block providing iops settings. See [IOPS settings](#iopsSettings) below for details.
+
+<a id="iopsSettings"></a>
+## IOPS settings
+
+(Supported from VCD *v10.2*)
+
+* `iops_limiting_enabled` - True if this storage profile is IOPs-based placement enabled
```
Comment (Didainius) [inline website/docs/d/storage_profile.html.markdown:52]:
```suggestion
* `iops_limiting_enabled` - True if this storage profile is IOPS-based placement enabled
```

File: website/docs/d/storage_profile.html.markdown [hunk 10]
```diff
@@ -32,5 +32,25 @@ The following arguments are supported:
...
 
 ## Attribute reference
+* `id` - storage profile identifier
 
-This data source exports only `id` field.
+(Supported in provider *v3.6+*)
+
+* `limit` - Maximum number of storage bytes (scaled by Units) allocated for this profile. A value of 0 is understood to mean `maximum possible`
+* `used_storage` - Storage used, in Megabytes, by the storage profile
+* `default` - True if this is default storage profile for this vDC. The default storage profile is used when an object that can specify a storage profile is created with no storage profile specified
```
Comment (Didainius) [inline website/docs/d/storage_profile.html.markdown:41]:
```suggestion
* `default` - True if this is default storage profile for this VDC. The default storage profile is used when an object that can specify a storage profile is created with no storage profile specified
```

File: website/docs/d/storage_profile.html.markdown [hunk 11]
```diff
@@ -32,5 +32,25 @@ The following arguments are supported:
 * `name` - (Required) Storage profile name.
 
 ## Attribute reference
+* `id` - storage profile identifier
 
-This data source exports only `id` field.
+(Supported in provider *v3.6+*)
+
+* `limit` - Maximum number of storage bytes (scaled by Units) allocated for this profile. A value of 0 is understood to mean `maximum possible`
+* `used_storage` - Storage used, in Megabytes, by the storage profile
```
Comment (Didainius) [inline website/docs/d/storage_profile.html.markdown:40]: Maybe it is just me, but it sounds strange - could this be simpler.

```suggestion
* `used_storage` - Storage used, by the storage profile (in Megabytes)
```

File: website/docs/d/storage_profile.html.markdown [hunk 12]
```diff
@@ -32,5 +32,25 @@ The following arguments are supported:
 * `name` - (Required) Storage profile name.
 
 ## Attribute reference
+* `id` - storage profile identifier
 
-This data source exports only `id` field.
+(Supported in provider *v3.6+*)
+
+* `limit` - Maximum number of storage bytes (scaled by Units) allocated for this profile. A value of 0 is understood to mean `maximum possible`
```
Comment (Didainius) [inline website/docs/d/storage_profile.html.markdown:39]:
```suggestion
* `limit` - Maximum number of storage bytes (scaled by 'units' field) allocated for this profile. A value of 0 is understood to mean `maximum possible`
```

File: .changes/v3.6.0/782-improvements.md (hunk 1 above)
Comment (vbauzysvmware) [reply .changes/v3.6.0/782-improvements.md]: changed

File: website/docs/d/storage_profile.html.markdown (hunk 12 above)
Comment (vbauzysvmware) [reply website/docs/d/storage_profile.html.markdown:39]: updated

File: website/docs/d/storage_profile.html.markdown (hunk 11 above)
Comment (vbauzysvmware) [reply website/docs/d/storage_profile.html.markdown:40]: changed

File: website/docs/d/storage_profile.html.markdown (hunk 8 above)
Comment (vbauzysvmware) [reply website/docs/d/storage_profile.html.markdown:54]: fixed

File: website/docs/d/storage_profile.html.markdown (hunk 7 above)
Comment (vbauzysvmware) [reply website/docs/d/storage_profile.html.markdown:55]: fixed

File: vcd/datasource_vcd_storage_profile.go (hunk 5 above)
Comment (vbauzysvmware) [reply vcd/datasource_vcd_storage_profile.go:68]: updated

File: vcd/datasource_vcd_storage_profile.go (hunk 4 above)
Comment (vbauzysvmware) [reply vcd/datasource_vcd_storage_profile.go:38]: updated

File: vcd/datasource_vcd_storage_profile.go [hunk 13]
```diff
@@ -32,6 +32,70 @@ func datasourceVcdStorageProfile() *schema.Resource {
...
+						},
+						"disk_iops_per_gb_max": &schema.Schema{
+							Type:        schema.TypeInt,
+							Computed:    true,
+							Description: "The maximum disk IOPs per GB value that this storage profile is permitted to deliver. A value of 0 means there is no per GB IOPS restriction",
+						},
+						"iops_limit": &schema.Schema{
+							Type:        schema.TypeInt,
+							Computed:    true,
+							Description: "Maximum number of IOPs that can be allocated for this profile. `0` means `maximum possible`",
```
Comment (Didainius) [inline vcd/datasource_vcd_storage_profile.go:94]:
```suggestion
							Description: "Maximum number of IOPS that can be allocated for this profile. `0` means `maximum possible`",
```

File: vcd/datasource_vcd_storage_profile.go [hunk 14]
```diff
@@ -32,6 +32,70 @@ func datasourceVcdStorageProfile() *schema.Resource {
...
+			},
+			"iops_allocated": &schema.Schema{
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "Total IOPS currently allocated to this storage profile",
+			},
+			"units": &schema.Schema{
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Scale used to define Limit",
```
Comment (Didainius) [inline vcd/datasource_vcd_storage_profile.go:63]:
```suggestion
				Description: "Scale used to define limit",
```

File: vcd/datasource_vcd_storage_profile.go [hunk 15]
```diff
@@ -32,6 +32,70 @@ func datasourceVcdStorageProfile() *schema.Resource {
...
 			},
+			"limit": &schema.Schema{
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "Maximum number of storage bytes (scaled by 'units' field) allocated for this profile. `0` means `maximum possible`",
+			},
+			"used_storage": &schema.Schema{
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "Storage used, in Megabytes, by the storage profile",
```
Comment (Didainius) [inline vcd/datasource_vcd_storage_profile.go:43]:
```suggestion
				Description: "Storage used, by the storage profile (in Megabytes)",
```

**Review by Didainius — APPROVED** [review]
A few more inconsistencies in docs, but the code itself worked in my manual testing

**Review by lvirbalas — APPROVED** [review]

**Review by mikeletux — APPROVED** [review]
LGTM!

**Pull request merged**